cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java

```diff
@@ -393,52 +393,86 @@ protected OpenScienceFrameworkCredential constructCredential(final RequestContex
-                logger.error("Notify Remote Principal Authenticated Exception: Institution Provider ID Required");
-                throw new RemoteUserFailedLoginException("Institution Provider ID Required");
-            }
-
-            if (username == null || username.trim().isEmpty()) {
-                logger.error("Notify Remote Principal Authenticated Exception: username=null, institution={}", institutionId);
-                throw new RemoteUserFailedLoginException("Username (Email) Required");
-            }
 
-            logger.info("Notify Remote Principal Authenticated: username={}, institution={}", username, institutionId);
-            logger.debug("Notify Remote Principal Authenticated [{}, {}] Normalized Payload '{}'", username, institutionId, payload);
+        // Step 1 - Normalize the remote principal
+        final JSONObject normalizedPayload;
+        try {
+            normalizedPayload = this.normalizeRemotePrincipal(credential);
+        } catch (final ParserConfigurationException | TransformerException e) {
+            logger.error("Failed to normalize remote principal: {}", e.getMessage());
+            logger.trace("Failed to normalize remote principal: {}", e);
+            throw new RemoteUserFailedLoginException("Failed to normalize remote principal.");
+        }
 
+        // Step 2 - Verify institution and user info in the normalized payload
+        final JSONObject provider = normalizedPayload.optJSONObject("provider");
+        if (provider == null) {
+            logger.error("Invalid remote principal: provider is required");
+            throw new RemoteUserFailedLoginException("Invalid remote principal: missing provider.");
+        }
+        final String institutionId = provider.optString("id").trim();
+        if (institutionId.isEmpty()) {
+            logger.error("Invalid remote principal: institution provider ID is required");
+            throw new RemoteUserFailedLoginException("Invalid remote principal: missing institution.");
+        }
+        final JSONObject user = provider.optJSONObject("user");
+        if (user == null) {
+            logger.error("Invalid remote principal: user is required");
+            throw new RemoteUserFailedLoginException("Invalid remote principal: missing user.");
+        }
+        final String username = user.optString("username").trim();
+        final String fullname = user.optString("fullname").trim();
+        final String givenName = user.optString("givenName").trim();
+        final String familyName = user.optString("familyName").trim();
+        if (username.isEmpty()) {
+            logger.error("Invalid remote principal: username (email) is required for institution '{}'", institutionId);
+            throw new RemoteUserFailedLoginException("Invalid remote principal: missing username.");
+        }
+        if (fullname.isEmpty() && (givenName.isEmpty() || familyName.isEmpty())) {
```

Comment:
Add a check for user names. Missing names are caught here in CAS instead of being delayed/delegated to OSF. After this update,  the only thing left for OSF to check is whether the email provided is valid and not blacklisted.